### DISCUSSION TITLE
Request for refresh token

### DISCUSSION THREAD

**Discussion by abelovq** [vercel/next.js#48864] [Help]
### Summary

Hello, maybe anyone knows, is there any way to implement refresh token mechanism (I mean if back-end returns 401, for example for get/users, we make a request for refresh token, if this request successful our interceptor re-request get/users one more time) for server components. Or is it just for client components ?

**Comment by kenneth-stylespace** [comment]
Did you mean like the OAuth2 flow? 
E.g., When client originally have a valid access token, then, the client makes an API request using the access token, but unfortunately, the access token is just expired. So, you want to get a NEW access token from the long-living refresh token. Then, run the original API request again (re-request) with the newly obtained access token.

If this is the case, this is definitely a client-side problem. You need to implement the an interceptor for the API request, i.e., when a 401 error is received during a request, send another request for getting a new access token by the refresh token, and then re-request the original request. BE VERY CAREFUL, the request for getting a new access token may also return 401 if your refresh token is invalid, then, it will be an infinite interceptor loop if you didn't handle the case probably.

If you are using `axios`, the interceptor is quite easy to implement. 
This link has an example: https://thedutchlab.com/blog/using-axios-interceptors-for-refreshing-your-api-token
The most important part is: 
```
const axiosApiInstance = axios.create();

// Response interceptor for API calls
axiosApiInstance.interceptors.response.use((response) => {
  return response
}, async function (error) {
  const originalRequest = error.config;
  if (error.response.status === 403 && !originalRequest._retry) {
    originalRequest._retry = true;
    const access_token = await refreshAccessToken();            
    axios.defaults.headers.common['Authorization'] = 'Bearer ' + access_token;
    return axiosApiInstance(originalRequest);
  }
  return Promise.reject(error);
});
```
As I am also using axios, I implemented my own one with Typescript and my own logic and code structure:
```
self.axios.interceptors.response.use((response) => {
  return response
}, async (error: AxiosError<{
  type: CustomHttpExceptionType
}> & {
  config: AxiosRequestConfig & {
    _retryForAccessToken: boolean,
    _retries: number,
  }
}) => {
  const originalRequest = error.config;
  if (originalRequest) {
    // 5 times retry
    originalRequest._retryForAccessToken = 
      (originalRequest._retryForAccessToken == undefined) ? true : originalRequest._retryForAccessToken;
      originalRequest._retries = originalRequest._retries ?? 5;

    if ((error?.response?.data?.type === "InvalidAccessTokenException") 
      && originalRequest._retryForAccessToken 
      && originalRequest._retries > 0
      ) {
      let refresh_token = self.getRefreshToken();
      if (refresh_token) {
        originalRequest._retries--;
        const {access_token} = await self.auth.refreshAccessToken({refresh_token}); 
        self.setAccessToken(access_token);
        return self.axios(originalRequest);
      }
      
    }
  }
  throw error;
});
```

If you are using `fetch`, you may implement this similarly with `fetch-retry`: https://www.npmjs.com/package/fetch-retry

And don't forget, no matter you are using `fetch` or `axios`, you must use the same `fetch` function or `axios` instance for all your existing ones whichever need the refresh token interceptor.

**Comment by abelovq** [comment]
Thank you @kenneth-stylespace for your answer, on client side I have almost the same you wrote, but my question is aimed for server componenets, 
according https://beta.nextjs.org/docs/data-fetching/fetching we can make get requests for getting smth, where we can use cookies() or headers(), but this realization does not assume refresh token flow, because we get cookies from browser, and if our access token have expired we make request /refresh (server update refresh token, response with new pair refresh and access) and if everything is OK, we repeat request, but method cookies() returns us old access token and it does not work, or am I missing smth ??

**Reply by kenneth-stylespace** [reply]
May I know your implementation as the refresh token flow (e.g., through API or what)?

**Reply by mattstyles** [reply]
You can do the same for server components as you do on the client, although maybe your question is really about how to get the refresh token on the server?

Cookies can be used to store the refresh token (and/or access token). When you get a 401, read the refresh token cookie, generate a new access token and use that to retry the request (if the refresh fails—i.e. invalid/expired refresh token—bail and redirect to login to get a new one).

You could also try to validate the access token in middleware and refresh it there, sending back the cookie. Note though that your server function is unable to see the refreshed token _until_ you reload the page and hit the url again (i.e. until the response has completed, triggering an update of cookies). I _think_ this is potentially a bug.
